File: src/analysis/NPart.cpp
/*
  Copyright (C) 2012,2013
      Max Planck Institute for Polymer Research
  Copyright (C) 2008,2009,2010,2011
      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI
  
  This file is part of ESPResSo++.
  
  ESPResSo++ is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
  
  ESPResSo++ is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>. 
*/

#include "python.hpp"
#include "NPart.hpp"
#include "storage/DomainDecomposition.hpp"

using namespace espressopp;

namespace espressopp {
  namespace analysis {
    real NPart::compute_real() const {

      int myN, systemN;
      System& system = getSystemRef();
      myN = system.storage->getNRealParticles();
      myN += system.storage->getNAdressParticles();
      boost::mpi::reduce(*getSystem()->comm, myN, systemN, std::plus<int>(), 0);
      
      return 1.0*systemN;
    }

    void NPart::registerPython() {
      using namespace espressopp::python;
      class_<NPart, bases< Observable > >
        ("analysis_NPart", init< shared_ptr< System > >())
      ;
    }
  }
}

File: src/analysis/NPart.py
#  Copyright (C) 2012,2013
#      Max Planck Institute for Polymer Research
#  Copyright (C) 2008,2009,2010,2011
#      Max-Planck-Institute for Polymer Research & Fraunhofer SCAI
#
#  This file is part of ESPResSo++.
#
#  ESPResSo++ is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  ESPResSo++ is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.


r"""
*************************
espressopp.analysis.NPart
*************************


.. function:: espressopp.analysis.NPart(system)

		:param system:
		:type system:
"""
from espressopp.esutil import cxxinit
from espressopp import pmi

from espressopp.analysis.Observable import *
from _espressopp import analysis_NPart

class NPartLocal(ObservableLocal, analysis_NPart):

    def __init__(self, system):
	if not (pmi._PMIComm and pmi._PMIComm.isActive()) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
            cxxinit(self, analysis_NPart, system)

if pmi.isController :
    class NPart(Observable, metaclass=pmi.Proxy):
        pmiproxydefs = dict(
            cls =  'espressopp.analysis.NPartLocal'
        )
